platform/mellanox/mlnx-platform-api/sonic_platform/sfp.py

```diff
@@ -351,44 +368,36 @@ def get_presence(self):
             bool: True if device is present, False if not
         """
         presence = False
-        ethtool_cmd = "ethtool -m sfp{} hex on offset 0 length 1 2>/dev/null".format(self.index)
         try:
-            proc = subprocess.Popen(ethtool_cmd, 
-                                    stdout=subprocess.PIPE, 
-                                    shell=True, 
-                                    stderr=subprocess.STDOUT, 
-                                    universal_newlines=True)
-            stdout = proc.communicate()[0]
-            proc.wait()
-            result = stdout.rstrip('\n')
-            if result != '':
+            with open((self.sdk_sysfs_page_path_header + PATH_PAGE00), mode='r+b', buffering=0) as f:
+                f.read(1)
                 presence = True
-
-        except OSError as e:
-            raise OSError("Cannot detect sfp")
+        except (OSError, IOError):
+            presence = False
 
         return presence
 
 
     # Read out any bytes from any offset
     def _read_eeprom_specific_bytes(self, offset, num_bytes):
-        eeprom_raw = []
-        ethtool_cmd = "ethtool -m sfp{} hex on offset {} length {}".format(self.index, offset, num_bytes)
         try:
-            output = subprocess.check_output(ethtool_cmd, 
-                                             shell=True, 
-                                             universal_newlines=True)
-            output_lines = output.splitlines()
-            first_line_raw = output_lines[0]
-            if "Offset" in first_line_raw:
-                for line in output_lines[2:]:
-                    line_split = line.split()
-                    eeprom_raw = eeprom_raw + line_split[1:]
-        except subprocess.CalledProcessError as e:
-            return None
-
-        return eeprom_raw
+            # if the data is on page0, offset indicate start of lower memory or upper memory
+            # if the data is on other page, then offset = page*256
+            page_num = int(offset/PAGE_SIZE) 
+            if page_num == 0:
+                page = self.sdk_sysfs_page_path_header + PATH_PAGE00
+            else:
+                if page_num == 1 and self.sfp_type == SFP_TYPE:
+                    page = self.sdk_sysfs_page_path_header + PATH_PAGE00_A2
+                else:
+                    page = self.sdk_sysfs_page_path_header + PATH_PAGE.format(page_num)
+                offset = offset - page_num * PAGE_SIZE
 
+            with open(page, mode='r+b', buffering=0) as f:
+                f.seek(offset)
+                return ["{:02x}".format(c) for c in f.read(num_bytes)]
```

Comment:
maybe just return f.read(num_bytes) ?

Reply:
the return value shall be a list like ['00', '06', 'ff', '00', '00', '00', '00', '00', '02', '00']
with f.read(num_bytes) will only get a string like b'\x03\x04#\x01\x00\x00\x04\x00\x04\x00'

Reply:
OK